design/external-authorization-design.md

```diff
@@ -0,0 +1,710 @@
+The GRPC protocol is a superset of the HTTP protocol and requires less configuration.
+The drawback of only supporting the GRPC protocol is that many existing Envoy authorization servers only support the HTTP protocol.
+Until Contour adds support for the v3 Envoy API, only the version 2 of the GRPC authorization protocol can be configured.
+
+Operators can bind an authorization service to a `HTTPProxy` using a new field in the `VirtualHost` struct.
+This field configures which `ExtensionService` resource to use, and allows the operator to set initial authorization policy.
+Authorization policy can also be set on a `Route`, so that application owners can pass metadata and disable authorization on specific routes.
+
+## Detailed Design
+
+### ExtensionService Changes
+
+The `ExtensionService` CRD has API version `projectcontour.io/v1alpha1`.
+
+There are a number of benefits to creating a CRD to represent a supporting service:
+
+- The CRD directly generates an Envoy [Cluster][4] that Contour can use for any purpose.
+  This means that with a single new API, Contour can add support for authorization, rate limiting and logging support services.
+- A CRD gives Contour a way to communicate the operational status of the support service, which was a [desired][9] [goal][7].
+- A CRD allows the team operating the support service and the team operating Contour to collaborate more loosely.
+
+```Go
+// SupportProtocolVersion is the version of the GRPC protocol used
+// to request support services.
+type SupportProtocolVersion string
+
+// SupportProtocolVersion2 requests the "v2" support protocol version.
+const SupportProtocolVersion2 SupportProtocolVersion = "v2"
+
+type ExtensionService struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ExtensionServiceSpec   `json:"spec,omitempty"`
+	Status ExtensionServiceStatus `json:"status,omitempty"`
+}
+
+// ExtensionServiceStatus should follow the pattern being established in [#2642][10].
+// This field will updated by Contour.
+type ExtensionServiceStatus struct {
+	Conditions []Condition `json:"conditions"`
+}
+
+type ExtensionServiceSpec struct {
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:Required
+	Services []contourv1.Service `json:"services"`
+
+	// The load balancing policy for sending requests to the service.
+	// +optional
+	LoadBalancerPolicy *contourv1.LoadBalancerPolicy `json:"loadBalancerPolicy,omitempty"`
+
+	// +optional
+	TimeoutPolicy *contourv1.TimeoutPolicy `json:"timeoutPolicy,omitempty"`
+
+	// This field sets the version of the GRPC protocol that Envoy uses to
+	// send requests to the support service. Since Contour always uses the
+	// v2 Envoy API, this is currently fixed at "v2". However, other
+	// protocol options will be available in future.
+	//
+	// +kubebuilder:validation:Enum=v2
+	ProtocolVersion SupportProtocolVersion `json:"protocolVersion,omitempty"`
```

Comment:
Question: is there timeplan for moving to Envoy API v3?  [Deprecation](https://www.envoyproxy.io/docs/envoy/latest/api/api_supported_versions#api-supported-versions) seems to be scheduled for end of the year.  

Comment: The `transport_api_version` parameters can only be set in [v3](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_authz/v3/ext_authz.proto#extensions-filters-http-ext-authz-v3-extauthz).  So user could only set the `protocolVersion` field after Contour has moved to v3.  I wonder if development of external authorization could be synchronized with the API version move so that v2 support could be skipped completely, and `protocolVersion` field in the spec could be removed?  If the Envoy v2 deprecation period holds, one could set `protocolVersion` to `v2` for only couple of months anyways.

Reply:
> Question: is there timeplan for moving to Envoy API v3?  [Deprecation](https://www.envoyproxy.io/docs/envoy/latest/api/api_supported_versions#api-supported-versions) seems to be scheduled for end of the year.  

There is an issue but it's not being worked on yet.

> 
> Comment: The `transport_api_version` parameters can only be set in [v3](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_authz/v3/ext_authz.proto#extensions-filters-http-ext-authz-v3-extauthz).  So user could only set the `protocolVersion` field after Contour has moved to v3.  I wonder if development of external authorization could be synchronized with the API version move so that v2 support could be skipped completely, and `protocolVersion` field in the spec could be removed?  If the Envoy v2 deprecation period holds, one could set `protocolVersion` to `v2` for only couple of months anyways.

In principle at least they are independent versions. I am pretty sure I asked htuch and he said they are likely to keep it, but I suppose that could always change. At any rate I'd prefer to not block on v3 API support.